File: Python/priorityQueue.py
from __future__ import with_statement #requies 2.5or above
from sys import argv
from threading import Lock
import math


class PriorityQueue:
    def __init__(self):
        self.queue_list = []
        self.pri_hash = {}

    def show(self):
        print self.queue_list

    def count(self):
        return len(self.queue_list)

    def priority_count(self):
        return len(self.pri_hash)

    def enqueue(self, pri, data):
        #Validate priority level to make sure it is positive integer
        if type(pri) is not int:
            raise ValueError("priority level must be a positive integer (non-integer given)")
        elif pri < 0:
            raise ValueError("priority level must be a positive integer (negative value given)")

        lock = Lock()
        queue_len = int(len(self.queue_list))
        current = queue_len
        parent = int(math.floor((current - 1) / 2))

        #Check to see if key for priority-level already exists in pri_hash dict
        #- if not- add one
        try:
            with lock:
                self.pri_hash[pri] += 1
        except KeyError:
            self.pri_hash[pri] = 1;

        with lock:
            self.queue_list.append({ 'pri' : pri, 'fifo_order' : self.pri_hash[pri], 'data' : data})

        #Bubble the new priority level up through the heap until it's parent is larger
        #or it's parent is equal but has a lower fifo_order
        while current > 0 and self.queue_list[current]['pri'] >= self.queue_list[parent]['pri']:
            if self.queue_list[current]['pri'] == self.queue_list[parent]['pri'] and self.queue_list[current]['fifo_order'] > self.queue_list[parent]['fifo_order']:
                break

            with lock:
                [ self.queue_list[current], self.queue_list[parent] ] = [ self.queue_list[parent], self.queue_list[current] ]
            current = parent
            parent = int(math.floor((current - 1) / 2))
        return self


    def dequeue(self):
        lock = Lock()
        q_length = len(self.queue_list)
        if q_length is 0:
            print "Queue is empty"
            return None

        returnData = self.queue_list[0]['data']
        #Decrement count for dequeued priority level in pri_hash dict
        with lock:
            self.pri_hash[ self.queue_list[0]['pri'] ] -= 1

        #If count reaches 0 delete the key for the priority-level
        if self.pri_hash[ self.queue_list[0]['pri'] ] == 0:
            self.pri_hash.pop( self.queue_list[0]['pri'], None )

        self.queue_list[0] = self.queue_list[q_length - 1]
        self.queue_list.pop()

        #Heapify list by priority-level first, then by fifo-order
        current = 0
        highest = 0
        left = 1
        right = 2
        in_position = False
        while current < q_length - 1 and not in_position:
            #Check to see whether current should swap with left or right child in heap
            try:
                if self.queue_list[current]['pri'] <= self.queue_list[left]['pri']:
                    if self.queue_list[current]['pri'] == self.queue_list[left]['pri'] and self.queue_list[current]['fifo_order'] > self.queue_list[left]['fifo_order']:
                        highest = left
                    if self.queue_list[current]['pri'] < self.queue_list[left]['pri']:
                        highest = left
            except IndexError:
                pass
            try:
                if self.queue_list[highest]['pri'] <= self.queue_list[right]['pri']:
                    if self.queue_list[highest]['pri'] == self.queue_list[right]['pri'] and self.queue_list[highest]['fifo_order'] > self.queue_list[right]['fifo_order']:
                        highest = right
                    if self.queue_list[highest]['pri'] < self.queue_list[right]['pri']:
                        highest = right
            except IndexError:
                pass

            if highest != current:
                with lock:
                    [ self.queue_list[current], self.queue_list[highest] ] = [ self.queue_list[highest], self.queue_list[current] ]
            else:
                in_position = True

            current = highest;
            left = current * 2 + 1
            right = current * 2 + 2

        return returnData

File: Javascript/priorityQueue.js
//Single linked list for holding values in FIFO order
class QueueNode{
  constructor(val){
    this.val = val;
    this.next = null;
  }
}

class FifoQueue{
  constructor(pri){
    this.pri = pri;
    this.head = null;
    this.tail = null;
  }

  enqueue(val){
    if(this.head === null){
      this.head = new QueueNode(val);
      this.tail = this.head;
    }
    else{
      this.tail.next = new QueueNode(val);
      this.tail = this.tail.next;
    }
    return this;
  }
  dequeue(){
    if(this.head === null){
      return null;
    }
    var returnVal = this.head.val;
    if(this.head === this.tail){
      this.head = null;
      this.tail = null;
    }
    else{
      this.head = this.head.next;
    }
    return returnVal;
  }
}

///////// End of FIFO Queue class

class PriorityQueue{
  constructor(){
    this.qCount = 0;
    this.queueArray = [];
    this.pCount = 0;
    this.indexMap = {};
  }
  count(){
    return this.qCount;
  }

  priorityCount(){
    return this.pCount;
  }


  enqueue(pri, data){
    if(arguments.length !== 2){
      throw {
        name : "Error",
        message : "Enqueue method accepts two arguments- (priority-level, data)",
      }
    }
    if(!Number.isInteger(pri) || pri < 0){
      throw {
        name : "Error",
        message : "Priority level must be a positive integer",
      }
    }
    this.qCount ++;
    var array = this.queueArray;
    var indexMap = this.indexMap;
    var queueLength = array.length;
    var current = queueLength;
    var parent = Math.floor((current - 1) / 2);

    //Check to see if a queue at this priority level has already been created
    if(indexMap[pri] === undefined){
      //If not- create a new FIFO queue and record its position in the indexMap
      this.pCount ++;
      array[queueLength] = new FifoQueue(pri).enqueue(data);
      indexMap[pri] = queueLength;

      //Bubble the new priority level up through the heap until it's parent larger
      while(current > 0 && array[current].pri > array[parent].pri){

        //Update the priorityQueue with the new index of each priority level
        indexMap[array[current].pri] = parent;
        indexMap[array[parent].pri] = current;

        //swap the current priority level with it's parent
        [ array[current], array[parent] ] = [ array[parent], array[current] ];

        //Re-set the current index to the new index and perform the check again
        current = parent;
        parent = Math.floor((current - 1) / 2);
      }
    }
    else{
      array[indexMap[pri]].enqueue(data);
    }
    return this;
  }

  dequeue(){
    var array = this.queueArray;

    if( array.length === 0 ){
      console.warn("Priority Queue is Empty");
      return undefined;
    }

    this.qCount --;
    var indexMap = this.indexMap;
    var queueLength = array.length;
    var returnData = array[0].dequeue();

    if(array[0].head === null){
      //If the queue at the top of the heap is empty, delete it's key from the indexMap and
      //decrement the priority count.
      this.pCount --;
      indexMap[array[0].pri] = null;
      delete indexMap[array[0].pri];
      if(Object.keys(indexMap).length === 0){
        //If the indexMap is empty our data structure is empty- return the last data;
        array.pop();
        return returnData;
      }
      //If the indexMap is not empty, move the last queue to the top
      indexMap[array[queueLength - 1].pri] = 0;
      array[0] = array.pop();

      var current = 0;
      var highest = current;
      var left = 1;
      var right = 2;
      var done = false;
      while(!done){
        //Check to see if either child has a higher priority
        if (left < array.length && array[left].pri > array[current].pri){
          highest = left;
        }
        if(right < array.length && array[right].pri > array[highest].pri){
          highest = right;
        }

        //If either child has a higher priority-
        if(highest !== current){
          //Update the priorityQueue with the new index of each priority level
          indexMap[array[current].pri] = highest;
          indexMap[array[highest].pri] = current;

          //Swap the current queue with whichever child has a higher priority level
          [ array[current], array[highest] ] = [ array[highest] , array[current] ];
        }
        //If neither child is larger than the current- the queue has reached it's correct position
        else{
          done = true;
        }

        //Reset the index markers and repeat
        current = highest;
        left = current * 2 + 1;
        right = current * 2 + 2;
      }
    }
    return returnData;
  }
}
